File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

vector<int> solution(int n) {
    vector<int> answer;
    int dirx[3] = {1,0,-1};
    int diry[3] = {0,1,-1};
    
    int target = 0;
    for(int i=1;i<=n;i++) target += i;
    
    vector<vector<int>> v(n, vector<int>(n,0));
    
    int num = 1;
    int x = 0;
    int y = 0;
    int dir = 0;
    while (num <= target) {
        v[x][y] = num;
        
        // 다음 좌표 계산
        int nx = x + dirx[dir];
        int ny = y + diry[dir];
        
        if (nx < 0 || nx >= n || ny < 0 || ny >= n || v[nx][ny] != 0) {
            dir = (dir + 1) % 3;
            nx = x + dirx[dir];
            ny = y + diry[dir];
        }
        
        x = nx;
        y = ny;
        num ++;
    }
    
    for(int i=0;i<n;i++) {
        for(int j=0;j<n;j++) {
            if (v[i][j] > 0) answer.push_back(v[i][j]);
        }
    }
    
    return answer;
}

File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.java
import java.util.*;

class Solution {
    public int[] solution(int n) {
        int target = 0;
        for(int i=1;i<=n;i++) target += i;
        
        int[] dirx = {1,0,-1};
        int[] diry = {0,1,-1};
        int[][] snail = new int[n][n];
        
        int num = 1;
        int d = 0;
        int x = 0, y = 0;
        while(num <= target) {
            snail[x][y] = num;
            
            int nx = x+dirx[d], ny = y+diry[d];
            if(nx<0 || nx>=n || ny<0 || ny>=n || snail[nx][ny] != 0) {
                d = (d+1) % 3;
                nx = x+dirx[d];
                ny = y+diry[d];
            }
            x = nx;
            y = ny;
            num++;
        }
        int[] answer = new int[target];
        int idx = 0;
        for(int[] s: snail) {
            for(int a: s) {
                if (a > 0) answer[idx++] = a;
            }
        }
        
        return answer;
    }
}